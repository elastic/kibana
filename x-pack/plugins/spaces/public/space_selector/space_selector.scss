.spcSpaceSelector {
  @include kibanaFullScreenGraphics;
}

.spcSpaceSelector__pageContent {
  background-color: transparent;
  box-shadow: none;
  border: none;
  text-align: center;
  position: relative;
  z-index: 10;
}

.spcSpaceSelector__heading {
  flex-direction: column;
  flex-shrink: 0;
  position: relative;
  padding: $euiSizeXL;
  z-index: 10;
}

.spcSpaceSelector__logo {
  @include kibanaCircleLogo;
  @include euiBottomShadowMedium;
  margin-bottom: $euiSizeXL;
}

.spcSpaceSelector__searchHolder {
  width: $euiFormMaxWidth; // make sure it's as wide as our default form element width
  max-width: 100%;
}
