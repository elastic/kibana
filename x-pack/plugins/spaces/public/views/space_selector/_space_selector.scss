@import '@elastic/eui/src/components/form/variables';

.spcSpaceSelector {
  padding: 0;
  position: fixed;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  z-index: $euiZLevel9 + 1000;
  background: inherit;
  background-image: linear-gradient(0deg, $euiColorLightestShade 0%, $euiColorEmptyShade 100%);
  overflow: auto;
}

.spcSpaceSelector::before {
  // SASSTODO: webpack pipeline isn't setup to handle image urls in SASS yet
  // content: url(../../assets/bg_top_branded.svg);
  position: absolute;
  top: 0;
  right: 0;
  z-index: 1;
}

.spcSpaceSelector::after {
  // SASSTODO: webpack pipeline isn't setup to handle image urls in SASS yet
  // content: url(../../assets/bg_bottom_branded.svg);
  position: fixed;
  bottom: -2px; // Hides an odd space at the bottom of the svg
  left: 0;
  z-index: 1;
}

.spcSpaceSelector__pageContent {
  background-color: transparent;
  box-shadow: none;
  border: none;
  text-align: center;
  position: relative;
  z-index: 10;
}

.spcSpaceSelector__heading {
  flex-direction: column;
  flex-shrink: 0;
  position: relative;
  padding: $euiSizeXL;
  z-index: 10;
}

.spcSpaceSelector__logo {
  margin-bottom: $euiSizeXL;
  @include kibanaCircleLogo;
  @include euiBottomShadowMedium;
}


.spcSpaceSelector__searchHolder {
  width: $euiFormMaxWidth; // make sure it's as wide as our default form element width
  max-width: 100%;
}
