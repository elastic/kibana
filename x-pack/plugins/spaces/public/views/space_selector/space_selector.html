<div ng-controller="spacesSelectorController" id="spaceSelectorRootWrap">
  <div id="spaceSelectorRoot" />
</div>
