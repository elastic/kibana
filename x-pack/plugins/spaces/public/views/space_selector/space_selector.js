/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React from 'react';
import {
  EuiPage,
  EuiPageHeader,
  EuiPageBody,
  EuiPageContent,
  EuiPageHeaderSection,
  EuiCard,
  EuiIcon,
  EuiText,
  EuiFlexGroup,
  EuiFlexItem,
} from '@elastic/eui';

export const SpaceSelector = (props) => (
  <EuiPage>
    <EuiPageHeader>
      <EuiPageHeaderSection className="logoHeader">
        <EuiIcon size="xxl" type={`logoKibana`} />
      </EuiPageHeaderSection>
    </EuiPageHeader>
    <EuiPageBody>
      <EuiPageContent>
        <EuiText className="spaceWelcomeText">
          <p className="welcomeLarge">Welcome to Kibana.</p>
          <p className="welcomeMedium">Select a space to begin.</p>
        </EuiText>

        <EuiFlexGroup gutterSize="l" justifyContent="spaceEvenly" className="spacesGroup">
          {props.spaces.map(renderSpace)}
        </EuiFlexGroup>

        <EuiText className="spaceProfileText">
          <p>You can change your workspace at anytime by accessing your profile within Kibana.</p>
        </EuiText>
      </EuiPageContent>
    </EuiPageBody>
  </EuiPage>
);

function renderSpace(space) {
  return (
    <EuiFlexItem key={space.id} grow={false} className="spaceCard">
      <EuiCard
        icon={<EuiIcon size="xxl" type={space.logo} />}
        title={renderSpaceTitle(space)}
        description={space.description}
        onClick={() => window.alert('Card clicked')}
      />
    </EuiFlexItem>
  );
}

function renderSpaceTitle(space) {
  return (
    <div className="spaceCardTitle">
      <span>{space.name}</span>
      {/* <EuiIcon type={'starEmpty'} size="s" /> */}
    </div>
  );
}
