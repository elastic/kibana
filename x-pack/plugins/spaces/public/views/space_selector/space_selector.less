.spcSpaceSelector::before {
  content: url(../../../assets/bg_top_branded.svg);
}

.spcSpaceSelector::after {
  content: url(../../../assets/bg_bottom_branded.svg);
}
