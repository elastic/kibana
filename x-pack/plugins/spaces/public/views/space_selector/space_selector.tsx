/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import {
  EuiFieldSearch,
  EuiFlexGroup,
  EuiFlexItem,
  EuiIcon,
  EuiPage,
  EuiPageBody,
  EuiPageContent,
  EuiPageHeader,
  EuiSpacer,
  EuiText,
  EuiTitle,
} from '@elastic/eui';
import { FormattedMessage, InjectedIntl, injectI18n } from '@kbn/i18n/react';
import { SpacesManager } from 'plugins/spaces/lib';
import React, { Component, Fragment } from 'react';
import { SPACE_SEARCH_COUNT_THRESHOLD } from '../../../common/constants';
import { Space } from '../../../common/model/space';
import { SpaceCards } from '../components/space_cards';

interface Props {
  spaces?: Space[];
  spacesManager: SpacesManager;
  intl: InjectedIntl;
}

interface State {
  loading: boolean;
  searchTerm: string;
  spaces: Space[];
}

class SpaceSelectorUI extends Component<Props, State> {
  constructor(props: Props) {
    super(props);

    const state: State = {
      loading: false,
      searchTerm: '',
      spaces: [],
    };

    if (Array.isArray(props.spaces)) {
      state.spaces = [...props.spaces];
    }

    this.state = state;
  }

  public componentDidMount() {
    if (this.state.spaces.length === 0) {
      this.loadSpaces();
    }
  }

  public loadSpaces() {
    this.setState({ loading: true });
    const { spacesManager } = this.props;

    spacesManager.getSpaces().then(spaces => {
      this.setState({
        loading: false,
        spaces,
      });
    });
  }

  public render() {
    const { spaces, searchTerm } = this.state;

    let filteredSpaces = spaces;
    if (searchTerm) {
      filteredSpaces = spaces.filter(
        space =>
          space.name.toLowerCase().indexOf(searchTerm) >= 0 ||
          (space.description || '').toLowerCase().indexOf(searchTerm) >= 0
      );
    }

    return (
      <EuiPage className="spcSpaceSelector" data-test-subj="kibanaSpaceSelector">
        <EuiPageBody>
          <EuiPageHeader className="spcSpaceSelector__heading">
            <EuiSpacer size="xxl" />
            <span className="spcSpaceSelector__logo">
              <EuiIcon size="xxl" type={`logoKibana`} />
            </span>

            <EuiTitle size="l">
              <h1>
                <FormattedMessage
                  id="xpack.spaces.spaceSelector.selectSpacesTitle"
                  defaultMessage="Select your space"
                />
              </h1>
            </EuiTitle>
            <EuiText size="s" color="subdued">
              <p>
                <FormattedMessage
                  id="xpack.spaces.spaceSelector.changeSpaceAnytimeAvailabilityText"
                  defaultMessage="You can change your space at anytime"
                />
              </p>
            </EuiText>
          </EuiPageHeader>
          <EuiPageContent className="spcSpaceSelector__pageContent">
            <EuiFlexGroup
              // @ts-ignore
              direction="column"
              alignItems="center"
              responsive={false}
            >
              {this.getSearchField()}
            </EuiFlexGroup>

            <EuiSpacer size="xl" />

            <SpaceCards spaces={filteredSpaces} onSpaceSelect={this.onSelectSpace} />

            {filteredSpaces.length === 0 && (
              <Fragment>
                <EuiSpacer />
                <EuiText
                  color="subdued"
                  // @ts-ignore
                  textAlign="center"
                >
                  <FormattedMessage
                    id="xpack.spaces.spaceSelector.noSpacesMatchSearchCriteriaDescription"
                    defaultMessage="No spaces match search criteria"
                  />
                </EuiText>
              </Fragment>
            )}
          </EuiPageContent>
        </EuiPageBody>
      </EuiPage>
    );
  }

  public getSearchField = () => {
    const { intl } = this.props;
    if (!this.props.spaces || this.props.spaces.length < SPACE_SEARCH_COUNT_THRESHOLD) {
      return null;
    }
    return (
      <EuiFlexItem className="spcSpaceSelector__searchHolder">
        <EuiFieldSearch
          className="spcSpaceSelector__searchField"
          placeholder={intl.formatMessage({
            id: 'xpack.spaces.spaceSelector.findSpacePlaceholder',
            defaultMessage: 'Find a space',
          })}
          incremental={true}
          // @ts-ignore
          onSearch={this.onSearch}
        />
      </EuiFlexItem>
    );
  };

  public onSearch = (searchTerm = '') => {
    this.setState({
      searchTerm: searchTerm.trim().toLowerCase(),
    });
  };

  public onSelectSpace = (space: Space) => {
    this.props.spacesManager.changeSelectedSpace(space);
  };
}

export const SpaceSelector = injectI18n(SpaceSelectorUI);
