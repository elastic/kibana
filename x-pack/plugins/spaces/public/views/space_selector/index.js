/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import 'ui/autoload/styles';
import chrome from 'ui/chrome';
import 'plugins/spaces/views/space_selector/space_selector.less';
import template from 'plugins/spaces/views/space_selector/space_selector.html';
import { SpacesManager } from 'plugins/spaces/lib/spaces_manager';
import { uiModules } from 'ui/modules';

import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { SpaceSelector } from './space_selector';

const module = uiModules.get('spaces_selector', []);
module.controller('spacesSelectorController', ($scope, $http, spaces) => {
  const domNode = document.getElementById('spaceSelectorRoot');

  const spacesManager = new SpacesManager($http, chrome);

  render(<SpaceSelector spaces={spaces} spacesManager={spacesManager} />, domNode);

  // unmount react on controller destroy
  $scope.$on('$destroy', () => {
    unmountComponentAtNode(domNode);
  });
});

chrome
  .setVisible(false)
  .setRootTemplate(template);
