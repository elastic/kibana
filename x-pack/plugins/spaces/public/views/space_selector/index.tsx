/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import { I18nProvider } from '@kbn/i18n/react';
import { SpacesManager } from 'plugins/spaces/lib/spaces_manager';
// @ts-ignore
import template from 'plugins/spaces/views/space_selector/space_selector.html';
import 'ui/autoload/styles';
import chrome from 'ui/chrome';
// @ts-ignore
import { uiModules } from 'ui/modules';

import 'plugins/spaces/views/space_selector/space_selector.less';
import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { Space } from '../../../common/model/space';
import { SpaceSelector } from './space_selector';

const module = uiModules.get('spaces_selector', []);
module.controller(
  'spacesSelectorController',
  ($scope: any, $http: any, spaces: Space[], spaceSelectorURL: string) => {
    const domNode = document.getElementById('spaceSelectorRoot');

    const spacesManager = new SpacesManager($http, chrome, spaceSelectorURL);

    render(
      <I18nProvider>
        <SpaceSelector spaces={spaces} spacesManager={spacesManager} />
      </I18nProvider>,
      domNode
    );

    // unmount react on controller destroy
    $scope.$on('$destroy', () => {
      if (domNode) {
        unmountComponentAtNode(domNode);
      }
    });
  }
);

chrome.setVisible(false).setRootTemplate(template);
