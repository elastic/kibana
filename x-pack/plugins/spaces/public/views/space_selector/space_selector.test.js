/*
 * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
 * or more contributor license agreements. Licensed under the Elastic License;
 * you may not use this file except in compliance with the Elastic License.
 */

import React from 'react';
import { SpaceSelector } from './space_selector';
import chrome from 'ui/chrome';
import renderer from 'react-test-renderer';
import { render, shallow } from 'enzyme';

function getHttpAgent(spaces = []) {
  const httpAgent = () => {};
  httpAgent.get = jest.fn(() => Promise.resolve({ data: spaces }));

  return httpAgent;
}


test('it renders without crashing', () => {
  const httpAgent = getHttpAgent();
  const component = renderer.create(
    <SpaceSelector spaces={[]} httpAgent={httpAgent} chrome={chrome}/>
  );
  expect(component).toMatchSnapshot();
});

test('it uses the spaces on props, when provided', () => {
  const httpAgent = getHttpAgent();

  const spaces = [{
    id: 'space-1',
    name: 'Space 1',
    description: 'This is the first space',
    urlContext: 'space-1-context'
  }];

  const component = render(
    <SpaceSelector spaces={spaces} httpAgent={httpAgent} chrome={chrome}/>
  );

  return Promise
    .resolve()
    .then(() => {
      expect(component.find('.spaceCard')).toHaveLength(1);
      expect(httpAgent.get).toHaveBeenCalledTimes(0);
    });
});

test('it queries for spaces when not provided on props', () => {
  const spaces = [{
    id: 'space-1',
    name: 'Space 1',
    description: 'This is the first space',
    urlContext: 'space-1-context'
  }];

  const httpAgent = getHttpAgent(spaces);

  const component = shallow(
    <SpaceSelector httpAgent={httpAgent} chrome={chrome}/>
  );

  return Promise
    .resolve()
    .then(() => {
      expect(httpAgent.get).toHaveBeenCalledTimes(1);
      expect(component.update().find('.spaceCard')).toHaveLength(1);
    });
});
