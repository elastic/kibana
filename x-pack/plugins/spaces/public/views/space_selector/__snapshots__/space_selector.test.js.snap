// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders without crashing 1`] = `
<div
  className="euiPage spaceSelector__page"
  style={undefined}
>
  <div
    className="euiPageBody"
    style={undefined}
  >
    <div
      className="euiPageHeader euiPageHeader--responsive spaceSelector__heading"
    >
      <div
        className="euiPageHeaderSection spaceSelector__logoHeader"
      >
        <div
          className="spaceSelector__logoCircle"
        >
          <svg
            className="euiIcon euiIcon--xxLarge"
            focusable="false"
            height="40"
            style={
              Object {
                "fill": undefined,
              }
            }
            tabIndex={undefined}
            viewBox="0 0 30 40"
            width="30"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M.874 40H30c0-8.046-4.566-15.197-11.652-19.768L.874 40z"
              fill="#3EBEB0"
            />
            <path
              d="M0 26v14h1.873L18.35 20.229a31.291 31.291 0 0 0-3.562-1.974L0 26z"
              fill="#37A595"
            />
            <path
              d="M0 13v23l14.79-17.748C10.424 16.186 15.38 14 10 14L0 13z"
              fill="#353535"
            />
            <path
              d="M30 0H0v15c5.38 0 10.424 1.186 14.79 3.252L30 0z"
              fill="#E9478B"
            />
          </svg>
        </div>
        <div
          className="euiSpacer euiSpacer--l"
        />
        <span
          className="euiTextColor euiTextColor--ghost euiTitle euiTitle--large"
        >
          <p>
            Select your space
          </p>
        </span>
      </div>
    </div>
    <div
      className="euiPanel euiPanel--paddingLarge euiPageContent spaceSelector__pageContent"
    >
      <div
        className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--alignItemsCenter euiFlexGroup--directionColumn"
      >
        <div
          className="euiFlexItem"
        >
          <div
            className="euiText euiText--extraSmall"
          >
            <p>
              You can change your space at anytime from within Kibana.
            </p>
          </div>
        </div>
      </div>
      <div
        className="euiSpacer euiSpacer--xl"
      />
      <div
        className="spaceCards"
      >
        <div
          className="euiFlexGroup euiFlexGroup--gutterLarge euiFlexGroup--justifyContentCenter euiFlexGroup--directionRow euiFlexGroup--wrap"
        />
      </div>
      <div
        className="euiSpacer euiSpacer--l"
      />
      <div
        className="euiText"
      >
        <div
          className="euiTextAlign euiTextAlign--center"
        >
          <div
            className="euiTextColor euiTextColor--subdued"
          >
            No spaces match search criteria
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
